File: src/test/java/com/github/nsc/de/shake/parser/ConstructorTests.java
package com.github.nsc.de.shake.parser;

import com.github.nsc.de.shake.parser.node.AccessDescriber;
import com.github.nsc.de.shake.parser.node.Tree;
import com.github.nsc.de.shake.parser.node.objects.ClassDeclarationNode;
import com.github.nsc.de.shake.parser.node.objects.ConstructorDeclarationNode;
import org.junit.jupiter.api.Test;

import static com.github.nsc.de.shake.TestUtil.assertType;
import static com.github.nsc.de.shake.parser.ParserTestUtil.parse;
import static org.junit.jupiter.api.Assertions.*;

public class ConstructorTests {

    @Test
    public void testConstructorType1() {

        Tree tree = parse("<ConstructorTest>", "class TestClass { constructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PACKAGE, node.getAccess());
        assertNull(node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testConstructorType1Children() {

        Tree tree = parse("<ConstructorChildrenTest>", "class TestClass { constructor() { print(10); } }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PACKAGE, node.getAccess());
        assertNull(node.getName());
        assertSame(1, node.getBody().getChildren().length);

    }

    @Test
    public void testPublicConstructorType1() {

        Tree tree = parse("<PublicConstructorTest>", "class TestClass { public constructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PUBLIC, node.getAccess());
        assertNull(node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testProtectedConstructorType1() {

        Tree tree = parse("<ProtectedConstructorTest>", "class TestClass { protected constructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PROTECTED, node.getAccess());
        assertNull(node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testPrivateConstructorType1() {

        Tree tree = parse("<PrivateConstructorTest>", "class TestClass { private constructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PRIVATE, node.getAccess());
        assertNull(node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testNamedConstructorType1() {

        Tree tree = parse("<NamedConstructorTest>", "class TestClass { constructor MyConstructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PACKAGE, node.getAccess());
        assertEquals("MyConstructor", node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testNamedConstructorType1Children() {

        Tree tree = parse("<NamedConstructorChildrenTest>", "class TestClass { constructor MyConstructor() { print(10); } }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PACKAGE, node.getAccess());
        assertEquals("MyConstructor", node.getName());
        assertSame(1, node.getBody().getChildren().length);

    }

    @Test
    public void testNamedPublicConstructorType1() {

        Tree tree = parse("<NamedPublicConstructorTest>", "class TestClass { public constructor MyConstructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PUBLIC, node.getAccess());
        assertEquals("MyConstructor", node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testNamedProtectedConstructorType1() {

        Tree tree = parse("<NamedProtectedConstructorTest>", "class TestClass { protected constructor MyConstructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PROTECTED, node.getAccess());
        assertEquals("MyConstructor", node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

    @Test
    public void testNamedPrivateConstructorType1() {

        Tree tree = parse("<NamedPrivateConstructorTest>", "class TestClass { private constructor MyConstructor() {} }");

        assertEquals(1, tree.getChildren().length);
        assertType(ClassDeclarationNode.class, tree.getChildren()[0]);

        ClassDeclarationNode classNode = (ClassDeclarationNode) tree.getChildren()[0];
        assertEquals("TestClass", classNode.getName());
        assertFalse(classNode.isFinal());
        assertFalse(classNode.isStatic());
        assertFalse(classNode.isInClass());
        assertSame(0, classNode.getClasses().length);
        assertSame(0, classNode.getFields().length);
        assertSame(0, classNode.getMethods().length);
        assertSame(1, classNode.getConstructors().length);

        ConstructorDeclarationNode node = classNode.getConstructors()[0];
        assertSame(AccessDescriber.PRIVATE, node.getAccess());
        assertEquals("MyConstructor", node.getName());
        assertSame(0, node.getBody().getChildren().length);

    }

}

File: src/test/kotlin/com/github/nsc/de/shake/parser/ConstructorTests.kt
package com.github.nsc.de.shake.parser

import com.github.nsc.de.shake.assertType
import org.junit.jupiter.api.Test
import com.github.nsc.de.shake.parser.node.objects.ClassDeclarationNode
import com.github.nsc.de.shake.parser.node.AccessDescriber
import org.junit.jupiter.api.Assertions

class ConstructorTests {
    @Test
    fun testConstructorType1() {
        val tree = ParserTestUtil.parse("<ConstructorTest>", "class TestClass { constructor() {} }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PACKAGE, node.access)
        Assertions.assertNull(node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testConstructorType1Children() {
        val tree = ParserTestUtil.parse("<ConstructorChildrenTest>", "class TestClass { constructor() { print(10); } }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PACKAGE, node.access)
        Assertions.assertNull(node.name)
        Assertions.assertSame(1, node.body.children.size)
    }

    @Test
    fun testPublicConstructorType1() {
        val tree = ParserTestUtil.parse("<PublicConstructorTest>", "class TestClass { public constructor() {} }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PUBLIC, node.access)
        Assertions.assertNull(node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testProtectedConstructorType1() {
        val tree = ParserTestUtil.parse("<ProtectedConstructorTest>", "class TestClass { protected constructor() {} }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PROTECTED, node.access)
        Assertions.assertNull(node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testPrivateConstructorType1() {
        val tree = ParserTestUtil.parse("<PrivateConstructorTest>", "class TestClass { private constructor() {} }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PRIVATE, node.access)
        Assertions.assertNull(node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testNamedConstructorType1() {
        val tree = ParserTestUtil.parse("<NamedConstructorTest>", "class TestClass { constructor MyConstructor() {} }")
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PACKAGE, node.access)
        Assertions.assertEquals("MyConstructor", node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testNamedConstructorType1Children() {
        val tree = ParserTestUtil.parse(
            "<NamedConstructorChildrenTest>",
            "class TestClass { constructor MyConstructor() { print(10); } }"
        )
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PACKAGE, node.access)
        Assertions.assertEquals("MyConstructor", node.name)
        Assertions.assertSame(1, node.body.children.size)
    }

    @Test
    fun testNamedPublicConstructorType1() {
        val tree = ParserTestUtil.parse(
            "<NamedPublicConstructorTest>",
            "class TestClass { public constructor MyConstructor() {} }"
        )
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PUBLIC, node.access)
        Assertions.assertEquals("MyConstructor", node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testNamedProtectedConstructorType1() {
        val tree = ParserTestUtil.parse(
            "<NamedProtectedConstructorTest>",
            "class TestClass { protected constructor MyConstructor() {} }"
        )
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PROTECTED, node.access)
        Assertions.assertEquals("MyConstructor", node.name)
        Assertions.assertSame(0, node.body.children.size)
    }

    @Test
    fun testNamedPrivateConstructorType1() {
        val tree = ParserTestUtil.parse(
            "<NamedPrivateConstructorTest>",
            "class TestClass { private constructor MyConstructor() {} }"
        )
        Assertions.assertEquals(1, tree.children.size)
        assertType(ClassDeclarationNode::class.java, tree.children[0])
        val classNode = tree.children[0] as ClassDeclarationNode
        Assertions.assertEquals("TestClass", classNode.name)
        Assertions.assertFalse(classNode.isFinal)
        Assertions.assertFalse(classNode.isStatic)
        Assertions.assertFalse(classNode.isInClass)
        Assertions.assertSame(0, classNode.classes.size)
        Assertions.assertSame(0, classNode.fields.size)
        Assertions.assertSame(0, classNode.methods.size)
        Assertions.assertSame(1, classNode.constructors.size)
        val node = classNode.constructors[0]
        Assertions.assertSame(AccessDescriber.PRIVATE, node.access)
        Assertions.assertEquals("MyConstructor", node.name)
        Assertions.assertSame(0, node.body.children.size)
    }
}